src/Internal/Payment/PaymentContext.php

```diff
@@ -7,9 +7,92 @@
 
 namespace WCPay\Internal\Payment;
 
+use Automattic\WooCommerce\Blocks\Payments\PaymentMethodTypeInterface;
+use WCPay\Internal\Payment\PaymentMethod\PaymentMethodInterface;
+
 /**
  * A context object, which is shared between payment states.
  */
 class PaymentContext {
+	/**
+	 * Contains all the context's data.
+	 *
+	 * @var array
+	 */
+	private $data = [];
+
+	/**
+	 * Stores an internal value.
+	 * Use this method for changes to allow logging in the future.
+	 *
+	 * @param string $key   Property name.
+	 * @param mixed  $value Value to store.
+	 */
+	private function set( string $key, $value ) {
+		$this->data[ $key ] = $value;
```

Comment:
Are we planning to add immutability to the certain properties? For example, you should not be able to change order_id once it is set, but with this approach, you can change it and cause wired cases during the execution of next states.

Reply:
I think the order ID is the only thing that might make sense as immutable, but everything else could change between requests. And the order ID is a specific case itself, as orders and payments should (probably?) have a 1:1 relationship. With that in mind, I'd put the order ID in the constructor, and remove the setter. WDYT?

Reply:
Done in https://github.com/Automattic/woocommerce-payments/pull/7324/commits/014155f907b6a09c53ef279c3baded1a446ec033. @zmaglica do you approve it this way?

Reply:
@htdat, in the last commit (you quoted it), I moved `$order_id` out of the `$data` array. So far, it looks like payments will belong to orders, and with that in mind, there should be a connection from the order to the payment, but not the other way around. This way, when we start saving the payment context, we would only care about serializing the `$data` array.

Reply:
>  I moved $order_id out of the $data array.

Aha, it's my oversight. If so, it looks good now IMO.